packages/salesforce/src/Adaptor.js

```diff
@@ -395,13 +321,117 @@ export function bulk(sObject, operation, options, fun) {
+  pollInterval: 6000,
+  pollTimeout: 240000,
+};
+/**
+ * Create and execute a bulk job using Salesforce Bulk API v2.
+ * @public
+ * @example
+ * bulk2(
+ *   "Patient__c",
+ *   "insert",
+ *   { failOnError: true, pollInterval: 3000, pollTimeout: 240000 },
+ *   (state) => {
+ *     return state.someArray.map((x) => ({ Age__c: x.age, Name: x.name }));
+ *   }
+ * );
+ * @function
+ * @param {String} sObject - API name of the sObject.
+ * @param {String} operation - The bulk operation to be performed
+ * @param {Object} options - Options passed to the bulk api.
+ * @param {Function} fun - A function which takes state and returns an array.
+ * @returns {Operation}
+ */
+export function bulk2(sObject, operation, options, fun) {
+  return async state => {
+    const { connection } = state;
+    const {
+      failOnError = true,
+      allowNoOp = true,
+      pollInterval = 3000,
+      pollTimeout = 9000,
+    } = options;
+
+    console.log(pollInterval, 'pollInterval');
+    console.log(pollTimeout, 'pollTimeout');
+
+    const finalAttrs = fun(state);
+
+    if (allowNoOp && finalAttrs.length === 0) {
+      console.info(
+        `No items in ${sObject} array. Skipping bulk ${operation} operation.`
+      );
+      return state;
+    }
+
+    if (finalAttrs.length > 10000) {
+      console.log('Your batch is bigger than 10,000 records; chunking...');
+    }
+
+    const chunkedBatches = chunk(finalAttrs, 10000);
+
+    let mapOptions = options;
+    if (mapOptions?.extIdField) {
+      const { extIdField, ...opts } = mapOptions;
+      mapOptions = { ...opts, externalIdFieldName: extIdField };
+    }
+
+    const results = await Promise.all(
+      chunkedBatches.map(async chunkedBatch => {
+        connection.bulk2.pollInterval = pollInterval;
+        connection.bulk2.pollTimeout = pollTimeout;
+
+        console.info(
+          `Creating bulk ${operation} job for ${sObject} with ${chunkedBatch.length} records`
+        );
+
+        await connection.bulk2
+          .loadAndWaitForResults({
+            object: sObject,
+            operation,
+            input: chunkedBatch,
+            ...mapOptions,
+          })
+          .then(res => {
+            const errors = res.failedResults;
+
+            errors.forEach(err => {
+              err[`${options.extIdField}`] =
+                chunkedBatch[err.position - 1][options.extIdField];
+            });
+
+            if (failOnError && errors.length > 0) {
+              console.error('Errors detected:');
+
+              throw new Error(JSON.stringify(errors, null, 2));
+            } else {
+              console.log('Successful Result : ', res.successfulResults.length);
+              console.log(
+                'Unprocessed Records : ',
+                res.unprocessedRecords.length
+              );
+              return res;
+            }
+          });
+      })
+    );
+
+    console.log('Merging results arrays.');
+    const merged = [].concat(...results);
```

Comment:
I am not sure what all this merging and concat'ing is trying to achieve!

Try dropping this line and just returning like this:
```
return { ...state, references: merged.concat(state.references) }
```